### PR TITLE
ISSUE-420 Implement JMAP Forward/Set

### PR DESCRIPTION
**Pull request by vttranlina**
Resolve https://github.com/linagora/tmail-backend/issues/420

### CODE REVIEW

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardSetMethod.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  override def getRequest(mailboxSession: MailboxSession, invocation: Invocation): Either[Exception, ForwardSetRequest] =
+    ForwardSetSerializer.deserializeForwardSetRequest(invocation.arguments.value) match {
+      case JsSuccess(forwardSetRequest, _) => Right(forwardSetRequest)
+      case errors: JsError => Left(new IllegalArgumentException(ResponseSerializer.serialize(errors).toString))
+    }
+
+  override def doProcess(capabilities: Set[CapabilityIdentifier], invocation: InvocationWithContext,
+                         mailboxSession: MailboxSession,
+                         request: ForwardSetRequest): Publisher[InvocationWithContext] =
+    SMono.fromCallable(() => request.parseUpdateRequest())
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardSetMethod.scala:43]: Does wrapping this in a mono make sense?

**Review by chibenwa — APPROVED** [review]

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraForwardSetMethodContract.scala [hunk 2]
```diff
@@ -0,0 +1,1028 @@
...
+      s"""{
+         |    "using": [ "urn:ietf:params:jmap:core",
+         |               "com:linagora:params:jmap:forward" ],
+         |    "methodCalls": [
+         |      ["Forward/set", {
+         |        "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |        "update": {
+         |            "singleton": {
+         |                "localCopy": true,
+         |                "forwards": []
```
Comment (Arsnael) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraForwardSetMethodContract.scala:883]: Your forwards are empty here, of course the owner would end up with the message... I would expect more a test with at least one forward and localCopy true and making sure it ends up in the owner mailbox instead here?

**Review by Arsnael — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardSetMethod.scala [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  private def update(mappingSource: MappingSource, forwardDestinations: Seq[MailAddress]): SMono[ForwardSetUpdateResult] =
+    retrieveMappings(mappingSource)
+      .map(currentForwards => (currentForwards.diff(forwardDestinations), forwardDestinations.diff(currentForwards)))
+      .flatMapMany {
+        case (deletedForwards, addedForwards) => deleteMappings(mappingSource, deletedForwards)
+          .thenMany(addMappings(mappingSource, addedForwards))
+      }.`then`()
+      .onErrorResume(error => SMono.just[ForwardSetUpdateResult](ForwardSetUpdateFailure("todo", error)))
```
Comment (quantranhong1999) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardSetMethod.scala:100]: It is a bit weird to see a "todo" here. However, it seems not to impact things.

**Comment by quantranhong1999** [comment]
Read it.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardSetMethod.scala (hunk 3 above)
Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardSetMethod.scala:100]: Fixed
https://github.com/linagora/tmail-backend/pull/430/files

**Pull request merged**